### PR TITLE
BZ#1796054 Managing infra machine sets

### PR DESCRIPTION
**Pull request by codyhoag**
https://bugzilla.redhat.com/show_bug.cgi?id=1796054

Preview: https://deploy-preview-28772--osdocs.netlify.app/openshift-enterprise/latest/machine_management/managing-infrastructure-machinesets.html

### CODE REVIEW

**Comment by codyhoag** [comment]
Preview: http://file.rdu.redhat.com/~choag/012221/infra-node-work/machine_management/managing-infrastructure-machinesets.html

**Comment by codyhoag** [comment]
@briantward do you mind taking a look at this first? This solely addresses the docs need in the Machine Management section; once we get the info squared away, a follow-up can be any necessary reorganizing in other areas.

Let me know if there's anything missing. Thanks!

**Comment by codyhoag** [comment]
@briantward bump. Do you have any suggestions for this before I begin the SME/QE review process?

**Comment by codyhoag** [comment]
@JoelSpeed not sure who would be best to provide an SME review for this. Can you review, or assign someone who may be better? Thanks!

**Comment by JoelSpeed** [comment]
@codyhoag I've reviewed the changes and they look ok to me, I'm not sure if there's a particular team that owns the concept of infra machines though so I'll ask around and see if anyone is particularly interested in taking a look at this

File: modules/creating-infra-machines.adoc [hunk 1]
```diff
@@ -3,10 +3,17 @@
...
-= Creating infrastructure machines
+= Creating infrastructure machines with a dedicated configuration
 
 If you need infrastructure machines to have dedicated configurations, then you must create an infra pool.
 
+[NOTE]
+====
+You do not need a custom pool unless you change a specific node to have a different set of machine configs applied to it. You do not need a custom machine config pool or machine config for infra nodes to work correctly, as it is not a strict requirement for tasks like node labeling, scheduling specific workloads, isolation, or preventing other workload scheduling.
+
+If you have altered a machine config, like removing a `worker` label from an infra node, you must create a custom pool to properly upgrade your node.
```
Comment (kikisdeliveryservice) [inline modules/creating-infra-machines.adoc:14]: This sentence doesn't quite make sense? You aren't altering a machine config you're removing a label - are you saying if you remove all labels from a an infra node, you should create a custom pool to ensure it gets upgraded?

File: modules/node-role-management.adoc [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// Module included in the following assemblies:
+//
+// * machine_management/managing-infrastructure-machinesets.adoc
+
+[id="node-role-management_{context}"]
+= Node role management
+
+By default, all nodes, excluding those on the control plane, are labeled with the `node-role.kubernetes.io/worker: ""` role. Nodes with just the `worker` role are called worker, or compute nodes. To create a specialized non-compute node that is not on the control plane, like an infrastructure or app node, it is recommended to keep the `worker` role and define the additional role too. For example, for an infra node, it should be labeled with the `node-role.kubernetes.io/worker: ""` and `node-role.kubernetes.io/infra: ""` roles. You can remove the `worker` role and only apply the specialized node role, but that requires additional steps for upgrading due to constraints with the Machine Config Operator (MCO).
```
Comment (kikisdeliveryservice) [inline modules/node-role-management.adoc:8]: Not sure what this last sentence means? It is allowed during upgrades to have custom pools no?
https://github.com/openshift/machine-config-operator/blob/master/docs/custom-pools.md#understanding-custom-pool-updates

File: modules/creating-infra-machines.adoc (hunk 1 above)
Comment (codyhoag) [reply modules/creating-infra-machines.adoc:14]: @kikisdeliveryservice I think the part stating `If you have altered a machine config` is misleading. From the context of this specific BZ I'm trying to solve (and based on my limited understanding), the removal of the `worker` label requires a custom pool to be created to properly upgrade the node.

Is that the only case that would require a custom pool? If so, I could change to something like:
>If you remove the `worker` label from an infra node, you must create a custom pool to properly upgrade your node.

File: modules/node-role-management.adoc (hunk 2 above)
Comment (codyhoag) [reply modules/node-role-management.adoc:8]: @kikisdeliveryservice The current issues that have been brought up in BZs are that customers want to strip the `worker` label and replace with an `infra` label. These doc updates are striving to recommend keeping the `worker` label, and also applying the `infra` label so they are both there. This scenario, as I've been told anyways, would not require a custom pool for upgrades.

What the last sentence is trying to say is that if you do remove the `worker` label, you'll need to take the additional steps to create the custom pool for upgrade purposes.

Can you confirm that is correct?

Comment (kikisdeliveryservice) [reply modules/node-role-management.adoc:8]: Yes, basically, a node needs to always be a member of some Machine Config Pool to be managed by the MCO. The MCO doesn't touch any nodes that aren't members of a MCP.

File: modules/creating-infra-machines.adoc (hunk 1 above)
Comment (kikisdeliveryservice) [reply modules/creating-infra-machines.adoc:14]: Gotcha - yes a node that is managed by the cluster/MCO needs to always be a member of a MCP.

**Comment by codyhoag** [comment]
@kikisdeliveryservice I made a couple edits based on your feedback. Let me know if there's anything else to address. Thanks!

File: modules/creating-infra-machines.adoc [hunk 3]
```diff
@@ -3,10 +3,17 @@
 // * post_installation_configuration/cluster-tasks.adoc
 
 [id="creating-infra-machines_{context}"]
-= Creating infrastructure machines
+= Creating infrastructure machines with a dedicated configuration
```
Comment (briantward) [inline modules/creating-infra-machines.adoc:6]: Why call it a "dedicated" configuration? It's no more dedicated than any other part I think.  This is just to describe how to use the MCP approach.

File: modules/creating-infra-machines.adoc [hunk 4]
```diff
@@ -3,10 +3,17 @@
...
 
 If you need infrastructure machines to have dedicated configurations, then you must create an infra pool.
 
+[NOTE]
+====
+You do not need a custom pool unless you change a specific node to have a different set of machine configs applied to it. You do not need a custom machine config pool or machine config for infra nodes to work correctly, as it is not a strict requirement for tasks like node labeling, scheduling specific workloads, isolation, or preventing other workload scheduling.
```
Comment (briantward) [inline modules/creating-infra-machines.adoc:12]: Maybe we can make this simpler?... Or more direct to the purpose?

"Optionally, if you would like to remove the worker label from your infrastructure nodes, you must create a MachineConfigPool specific to your infrastructure nodes."

Or, with an explanation to the above?

"Optionally, if you would like to remove the worker label from your infrastructure nodes, you must create a MachineConfigPool specific to your infrastructure nodes, as the worker MCP finds members of its pool by the worker role.  If you remove the role that defines that membership from a node, you must create a new MCP for that node to continued to be controlled by the MCO."

File: modules/node-role-management.adoc [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+You do not need a custom pool unless a specific node needs to be updated to have a different set of machine configs applied to it. You do not need a custom machine config pool or machine config for infrastructure nodes to work correctly, as it is not a strict requirement for tasks like node labeling, scheduling specific workloads, isolation, and preventing workload scheduling. You can, however, use a custom machine config pool if you find it useful for organizing and managing your machine configs. Visit the "Creating infrastructure machines with a dedicated configuration" section for more information.
```
Comment (briantward) [inline modules/node-role-management.adoc:10]: I favor removing this in exchange for the simpler statement "Optionally, if you would like to remove the worker label from your infrastructure nodes, you must create a MachineConfigPool specific to your infrastructure nodes."

File: modules/node-role-management.adoc [hunk 6]
```diff
@@ -0,0 +1,14 @@
+// Module included in the following assemblies:
+//
+// * machine_management/managing-infrastructure-machinesets.adoc
+
+[id="node-role-management_{context}"]
+= Node role management
+
+By default, all nodes, excluding those on the control plane, are labeled with the `node-role.kubernetes.io/worker: ""` role. Nodes with just the `worker` role are called worker, or compute nodes. To create a specialized non-compute node that is not on the control plane, like an infrastructure or app node, it is recommended to keep the `worker` role and define the additional role too. For example, for an infra node, it should be labeled with the `node-role.kubernetes.io/worker: ""` and `node-role.kubernetes.io/infra: ""` roles. You can remove the `worker` role and only apply the specialized node role, but that requires creating a custom pool to allow for node upgrades due to constraints with the Machine Config Operator (MCO).
```
Comment (briantward) [inline modules/node-role-management.adoc:8]: I think this is the point under contention: "it is recommended to keep the `worker` role".  As @kikisdeliveryservice has pointed out, the MCO does support the ability to remove the worker role now, whereas a while back that was not the case.  
Also, just adding the label "infra" doesn't really make it a specialized non-compute node.  You've still got to either apply taints and tolerations or change your default node selector in your scheduler. I believe you already stated that above.  I might go so far as to say this whole subsection "node role management" could be removed.

**Review by briantward — COMMENTED** [review]
While not in this particular commit, I also notice that you are showing how to change the infra components just using the nodeselector, but you haven't shown how to prevent customer workloads from being scheduled on infra (i.e. add taints and tolerations OR change the default node selector in the scheduler)

File: machine_management/managing-infrastructure-machinesets.adoc [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+include::modules/infrastructure-components.adoc[leveloffset=+1]
+
+== Workload placement for infra nodes
+
+The two primary reasons for infra nodes include:
+
+* Preventing incurring billing costs against subscription counts.
+* Separating maintenance workloads and management workloads.
+
+After creating an infrastructure machine set, the new hosted infrastructure nodes have both the `worker` and `infra` roles applied to them. The `worker` role is required when the node is created, and is also used during node upgrades. Nodes with the `infra` role applied are not counted toward the total number of subscriptions that are required to run the environment, even when the `worker` role is also applied. The `worker` role can be removed, but it requires additional procedures to reconfigure the associated machine config pool to manage node upgrades.
```
Comment (sferich888) [inline machine_management/managing-infrastructure-machinesets.adoc:19]: The core issue with this is that, infra component and 'customer workloads' can by default run on 'infra' nodes. 

Some customers may not want this! We need to elaborate more in this section (or provide a link to instructions or a set of materials) that outline how to isolate infra nodes to infra components, only.

File: machine_management/managing-infrastructure-machinesets.adoc [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+====
+Unlike earlier versions of {product-title}, you cannot move the infrastructure components to the control plane machines. To move the components, you must create a new machine set.
+====
+
+With an infra node being assigned as a worker, there is a chance user workloads could get inadvertently assigned to an infra node. To prevent other workloads from being scheduled onto infrastructure nodes, there are two possible workflows:
+
+* Apply a taint to the infrastructure nodes and tolerations to the desired infrastructure workloads.
+* Apply a completely separate label to your other nodes and matching node selector to your other workloads such that they are mutually exclusive from infrastructure nodes.
+
+This choice mostly depends on the infrastructure components you plan to assign to the infra nodes. Several infrastructure components do not support configuring taints and tolerations; in those cases, you must apply a separate label to your other nodes so they are mutually exclusive from infrastructure nodes.
```
Comment (sferich888) [inline machine_management/managing-infrastructure-machinesets.adoc:31]: We should document which 'infra components' don't support this. The list of infra components is not 'long'. Our subscription guides outline which components those are clearly: 

- Red Hat OpenShift-included registry
- Router
- OpenShift cluster monitoring
- OpenShift log aggregation
- Red Hat Quay
- Red Hat OpenShift Container Storage
- Red Hat Advanced Cluster Manager

So which of these do you have to use the separate label situation with? 
Honestly if 'one' of these doesn't support 'taints and tolerations' we should probably be promoting/using only the labeling method.

File: machine_management/managing-infrastructure-machinesets.adoc [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+include::modules/node-role-management.adoc[leveloffset=+1]
+
+[id="assigning-machineset-resources-to-infra-nodes"]
+== Ensuring resources are scheduled to infrastructure nodes
+
+You can ensure only infrastructure workloads are scheduled to infra nodes by applying a taint to the infrastructure nodes and tolerations to the desired infrastructure workloads. If you prefer to avoid taints, this is also possible through a specified scheduling behavior. You can schedule your own workloads to run on specific non-infrastructure nodes by applying a distinct label to the other nodes and either updating your default scheduler to choose this particular node label or specifically request this node label by project namespace.
+
+include::modules/binding-infra-node-workloads-using-taints-tolerations.adoc[leveloffset=+2]
```
Comment (sferich888) [inline machine_management/managing-infrastructure-machinesets.adoc:40]: Depending on what in in this module, we may want to elaborate more in this section on what to do, explicitly. 

As you say above this doesn't work for all 'components' and as such using this method should be considered carefully (secondary to using a label). 

We don't seem to document this (as far as I can tell), so have we focused on the wrong 'suggestion'?

**Review by sferich888 — CHANGES_REQUESTED** [review]
I have a lot of reservations about this, and feel that more followup in https://issues.redhat.com/browse/RFE-1136 is needed on this.

**Comment by codyhoag** [comment]
> I have a lot of reservations about this, and feel that more followup in https://issues.redhat.com/browse/RFE-1136 is needed on this.

@sferich888 I've responded in the RFE with more detail, but in summary, I 100% agree! I will hold on this until there is a concrete plan on how to properly document this. Thanks!

File: modules/node-role-management.adoc (hunk 6 above)
Comment (kikisdeliveryservice) [reply modules/node-role-management.adoc:8]: Just to reiterate @briantward point about OCP doesn't really have a concept of infra nodes - this type of documentation is more a customer courtesy - an infra pool is just another custom pool from the perspective of the MCO that must be further configured as necessary to reach whatever specific goals you have for it.

**Comment by codyhoag** [comment]
The docs for managing infra workloads have been adjusted and resubmitted in #30683.

**Pull request closed without merging**